pkg/workflows/wasm/log_codec.go

```diff
@@ -0,0 +1,159 @@
+package wasm
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"math"
+	"reflect"
+	"sync"
+	"time"
+
+	"go.uber.org/zap/buffer"
+	"go.uber.org/zap/zapcore"
+)
+
+type logCodec struct {
+	zapcore.ObjectEncoder
+	sync.Mutex
+}
+
+type completeLog struct {
+	Entry  zapcore.Entry   `json:"entry"`
+	Fields []zapcore.Field `json:"fields"`
+}
+
+func NewLogCodec() *logCodec {
+	return &logCodec{}
+}
+
+func (c *logCodec) Clone() zapcore.Encoder {
+	c.Lock()
+	defer c.Unlock()
+
+	return &logCodec{}
+}
+
+func (c *logCodec) EncodeEntry(entry zapcore.Entry, fields []zapcore.Field) (*buffer.Buffer, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	cl := completeLog{
+		Entry:  entry,
+		Fields: fields,
+	}
+
+	logBytes, err := json.Marshal(cl)
+	if err != nil {
+		return nil, err
+	}
+
+	buf := buffer.NewPool().Get()
+	buf.AppendBytes(logBytes)
+
+	return buf, nil
+}
+
+func (c *logCodec) Decode(b []byte) (*zapcore.Entry, []interface{}, error) {
+	c.Lock()
+	defer c.Unlock()
```

Comment:
Why lock? Isn't this a pure implementation that doesn't use `c`?

Reply:
I missed this in my previous review :/ Yeah we don't need to lock here -- the implementation is pure + WASM is single-threaded.

Reply:
thanks for pointing this, its a left over. initially I had the buffer as part of the state. will remove it thanks!